### ISSUE TITLE
GPU memory consumption increases over the Epoch a lot

### ISSUE THREAD

**Issue by ManuelM95** [IDEA-Research/detrex#149]
Hi, 

thanks for the great work. I successful trained the 4-scale-swin-L model on a smaller Dataset with a few 1000 images which are 1024x1024 each. When training now with 100k images (same size as before) per epoch the GPU memory seems to slowly increase until the GPU goes out of memory. Is that the normal behavior? I train with batchsize 1 per GPU on TITAN RTX 24GB GPUs.

Thinking it might be the bigger serialized dataset which takes 435.86 MiB - I went from a Swin -L to a Swin-B backbone with the same result just after more iterations. Starting with around 12GB being allocated at the start - it goes up until it doesn't fit the 24GB in ~20k iterations. I attached the log of that training with 1 GPU and batch size 1.

Support would be highly appreciated :) 

[example_log.LOG](https://github.com/IDEA-Research/detrex/files/10114806/example_log.LOG)

**Comment by SlongLiu** [comment]
Thanks for providing the issue. 

### 1. Common phenomenon if there is a mild mem increase in early epochs.
We observed the increasing memory usage during training in both the original DINO and the detrex version,  but only in the early epochs and most mild. 

Some reasons might contribute to the phenomenon: 1) DINO random resizes images as augmentation. Early training samples might have small sizes. The max image size will be (1333, 800) if bs=1. If you want to know the max GPU mem needed, you might run a test script, where all training images are resized to the max sizes. 2) The GPU mem needed is related to the ground truth boxes annotated in your training images. Some training images might have more annotated boxes, which results in a mem peak after training for a while. 

In summary, it is common to have a mild memory increase in the early steps.

### 2. For you case
I looked at your log and found the GPU utilization increased significantly, even after 10 epochs, which is strange to me. We checked our training logs on COCO. The GPU mem usage increases from about 32G (~1k steps) to 38G (~60k steps), and the usage become stable then. I speculate it to the uncenterty mentioned above. For example, a image with many annotations have the max random sizes at a training point, which will cause a mem peak.

**Comment by ManuelM95** [comment]
Thanks for the answer:

I checked the amount of annotation per image: They vary a lot and can go up to 6.5k annotated objects in one patch. So that's probably the cause of the memory peaks. We hopefully soon get new GPUs - so I can test DINO with the data.

Or do you see any ideas of avoiding the peaks?

**Comment by SlongLiu** [comment]
There are two ways to reduce GPU mems. The first is to use smaller models or input resolution. Another is to introduce some engineering techniques. Two commonly used ways are mixed precision training and gradient checkpointing.

**Comment by ManuelM95** [comment]
Thanks for the advice - I will try it out :) .
